File: binding/scaxls.rb
class Scaxls
  def self.type_and_value(val)
    if val.instance_of?(String) or val.is_a?(Numeric)
      return nil, val
    elsif val.instance_of?(Time) or val.instance_of?(DateTime)
      return "datetime", val.to_f * 1000.0
    elsif val.instance_of?(Date)
      return "date", val.to_time.to_f * 1000.0
    elsif val.instance_of?(TrueClass) or val.instance_of?(FalseClass)
      return nil, val.to_s
    else
      return nil, val
    end
  end

  def self.autoload
    here = File.expand_path(File.dirname(__FILE__))
    lib = Dir.glob(File.join(here, "scaxls-*.jar")).sort.last
    raise RuntimeException.new("library not found.") if lib.nil?
    lib
  en

  def self.json2xls(json_data, lib=autoload)
    IO.popen("java -jar #{lib}", "w+b") do |io|
      io.write(json_data)
      io.close_write
      io.read
    end
  end

  def self.xls2json(filename, lib=autoload)
    IO.popen("java -jar #{lib} -template #{filename} -mode read", "r") {|io| io.read}
  end
end

File: binding/scaxls.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-

import os
import glob
import time
import subprocess

from datetime import datetime, date

def type_and_value(val):
    _type = type(val)
    if _type in [str, int, long, float]:
        return None, val
    elif _type is datetime:
        return "datetime", time.mktime(val.timetuple())
    elif _type is date:
        return "date", time.mktime(val.timetuple())
    elif _type is bool:
        return None, str(val)
    else:
        return None, val

def autoload():
    here = os.path.abspath(os.path.dirname(__file__))

    suggestions = glob.glob(os.path.join(here, "scaxls-*.jar"))
    if len(suggestions) == 0:
        raise RuntimeError("library not found.")

    return sorted(suggestions, reverse=True)[0]

def json2xls(json, libloader=autoload):
    proc = subprocess.Popen("java -jar {0}".format(libloader()),
                            shell=True,
                            stdin=subprocess.PIPE,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE)

    out, err = proc.communicate(json)
    return out

def xls2json(filename, libloader=autoload):
    proc = subprocess.Popen("java -jar {0} -template {1} -mode read".format(libloader(), filename),
                            shell=True,
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE)
    out, err = proc.communicate()
    return out

if __name__ == '__main__':
    import sys
    #print xls2json(sys.argv[1])

    print json2xls(sys.stdin.read())
